### ISSUE TITLE
Constructing deep objects in macros causes segfaults or weird syntax errors

### ISSUE THREAD

**Issue by neoeno** [oven-sh/bun#11730]
### How can we reproduce the crash?

The tl;dr is that constructing large-ish/complex objects within a function that is called as part of a [macro](https://bun.sh/docs/bundler/macros) causes intermittent segfaults, odd syntactical errors, or will occasionally just work.

Minimal reproduction here: https://github.com/neoeno/bun-macros-crash-repro

The original use-case that triggered this bug: I was playing around with the macros feature to see if I could use it to precompile markdown into jsx in my build process, which then led to me using [UnifiedJS](http://unifiedjs.com) to compile MDX to an AST, returning this from the macro, and hoping I could then convert it back to React nodes on the server. This is definitely not sensible but it did lead to this interesting bug :)

The behaviour of the bug is interesting — sometimes there are segfaults and sometimes part of the object seems to get into the wrong place and get executed weirdly (see the last example which reads `SyntaxError: Unexpected identifier 'YAY'. Expected a ':' following the property name 'static'.` where one of the object keys seems to get into the wrong place).

While probably no one is doing this in the real world, all of this seemed 'strange' enough that I thought better to be safe and report it in case it points to any larger issues. Hopefully not! In any case thanks to for your work on Bun.

### Relevant log output

```shell
# Since the bug causes intermittent results, I've run it a number of times here so you can see the general distribution of the different output.
$ bun --revision  
1.1.12+43f0913c3
$ uname -mprs     
Darwin 23.5.0 arm64 arm
$ bun run index.ts
1 | import { macroified } from './macro' with { type: "macro" };
2 | console.log(macroified());
    ^
SyntaxError: Unexpected identifier 'YAY'. Expected a ':' following the property name 'static'.
      at <parse> (/Users/kay/Code/tmp/macrosrepro/index.ts:2:1)

Bun v1.1.12 (macOS arm64)
$ bun run index.ts
============================================================
Bun v1.1.12 (43f0913c) macOS Silicon
Args: "bun" "run" "index.ts"
Features: jsc macros(2) tsconfig
Builtins: "bun:main"
Elapsed: 13ms | User: 11ms | Sys: 8ms
RSS: 23.53MB | Peak: 23.53MB | Commit: 1.07GB | Faults: 90

panic(main thread): Segmentation fault at address 0x29
oh no: Bun has crashed. This indicates a bug in Bun, not your code.

To send a redacted crash report to Bun's team,
please file a GitHub issue using the link below:

 https://bun.report/1.1.12/Mr143f0913AiggghE__upk5Tupk5T+lg5Tupk5T2t55T+/t5T+rysR+5trQA2AyC

zsh: trace trap  bun run index.ts
$ bun run index.ts
============================================================
Bun v1.1.12 (43f0913c) macOS Silicon
Args: "bun" "run" "index.ts"
Features: jsc macros(2) tsconfig
Builtins: "bun:main"
Elapsed: 14ms | User: 12ms | Sys: 9ms
RSS: 23.63MB | Peak: 23.63MB | Commit: 1.07GB | Faults: 46

panic(main thread): Segmentation fault at address 0x29
oh no: Bun has crashed. This indicates a bug in Bun, not your code.

To send a redacted crash report to Bun's team,
please file a GitHub issue using the link below:

 https://bun.report/1.1.12/Mr143f0913AiggghE__upk5Tupk5T+lg5Tupk5T2t55T+/t5T+rysR+5trQA2AyC

zsh: trace trap  bun run index.ts
$ bun run index.ts
============================================================
Bun v1.1.12 (43f0913c) macOS Silicon
Args: "bun" "run" "index.ts"
Features: jsc macros(2) tsconfig
Builtins: "bun:main"
Elapsed: 11ms | User: 12ms | Sys: 7ms
RSS: 23.64MB | Peak: 23.64MB | Commit: 1.07GB | Faults: 45

panic(main thread): Segmentation fault at address 0x29
oh no: Bun has crashed. This indicates a bug in Bun, not your code.

To send a redacted crash report to Bun's team,
please file a GitHub issue using the link below:

 https://bun.report/1.1.12/Mr143f0913AiggghE__upk5Tupk5T+lg5Tupk5T2t55T+/t5T+rysR+5trQA2AyC

zsh: trace trap  bun run index.ts
$ bun run index.ts
============================================================
Bun v1.1.12 (43f0913c) macOS Silicon
Args: "bun" "run" "index.ts"
Features: jsc macros(2) tsconfig
Builtins: "bun:main"
Elapsed: 13ms | User: 12ms | Sys: 9ms
RSS: 23.43MB | Peak: 23.43MB | Commit: 1.07GB | Faults: 45

panic(main thread): Segmentation fault at address 0x29
oh no: Bun has crashed. This indicates a bug in Bun, not your code.

To send a redacted crash report to Bun's team,
please file a GitHub issue using the link below:

 https://bun.report/1.1.12/Mr143f0913AiggghE__upk5Tupk5T+lg5Tupk5T2t55T+/t5T+rysR+5trQA2AyC

zsh: trace trap  bun run index.ts
$ bun run index.ts
============================================================
Bun v1.1.12 (43f0913c) macOS Silicon
Args: "bun" "run" "index.ts"
Features: jsc macros(2) tsconfig
Builtins: "bun:main"
Elapsed: 7ms | User: 7ms | Sys: 5ms
RSS: 23.48MB | Peak: 23.48MB | Commit: 1.07GB | Faults: 55

panic(main thread): Segmentation fault at address 0x2020202020202048
oh no: Bun has crashed. This indicates a bug in Bun, not your code.

To send a redacted crash report to Bun's team,
please file a GitHub issue using the link below:

 https://bun.report/1.1.12/Mr143f0913AiggghE__upk5Tupk5T+lg5Tupk5T2t55T+/t5T+rysR+5trQA2giwgkgBwkwgkgB

zsh: trace trap  bun run index.ts
$ bun run index.ts
============================================================
Bun v1.1.12 (43f0913c) macOS Silicon
Args: "bun" "run" "index.ts"
Features: jsc macros(2) tsconfig
Builtins: "bun:main"
Elapsed: 16ms | User: 11ms | Sys: 8ms
RSS: 41.68MB | Peak: 41.68MB | Commit: 1.07GB | Faults: 89

panic(main thread): Segmentation fault at address 0x16E8C36A8
oh no: Bun has crashed. This indicates a bug in Bun, not your code.

To send a redacted crash report to Bun's team,
please file a GitHub issue using the link below:

 https://bun.report/1.1.12/Mr143f0913AiggghE__2hk5T+lg5Tupk5T+lg5Tupk5Tupk5Tupk5T+lg5TA2Cwq7wxuD

zsh: trace trap  bun run index.ts
$ bun run index.ts
============================================================
Bun v1.1.12 (43f0913c) macOS Silicon
Args: "bun" "run" "index.ts"
Features: jsc macros(2) tsconfig
Builtins: "bun:main"
Elapsed: 14ms | User: 12ms | Sys: 10ms
RSS: 23.79MB | Peak: 23.79MB | Commit: 1.07GB | Faults: 80

panic(main thread): Segmentation fault at address 0x29
oh no: Bun has crashed. This indicates a bug in Bun, not your code.

To send a redacted crash report to Bun's team,
please file a GitHub issue using the link below:

 https://bun.report/1.1.12/Mr143f0913AiggghE__upk5Tupk5T+lg5Tupk5T2t55T+/t5T+rysR+5trQA2AyC

zsh: trace trap  bun run index.ts
$ bun run index.ts
============================================================
Bun v1.1.12 (43f0913c) macOS Silicon
Args: "bun" "run" "index.ts"
Features: jsc macros(2) tsconfig
Builtins: "bun:main"
Elapsed: 17ms | User: 15ms | Sys: 11ms
RSS: 41.44MB | Peak: 41.44MB | Commit: 1.07GB | Faults: 46

panic(main thread): Segmentation fault at address 0x16DF536A8
oh no: Bun has crashed. This indicates a bug in Bun, not your code.

To send a redacted crash report to Bun's team,
please file a GitHub issue using the link below:

 https://bun.report/1.1.12/Mr143f0913AiggghE__2hk5T+lg5Tupk5T+lg5Tupk5Tupk5Tupk5T+lg5TA2Cwq70+tD

zsh: trace trap  bun run index.ts
$ bun run index.ts
============================================================
Bun v1.1.12 (43f0913c) macOS Silicon
Args: "bun" "run" "index.ts"
Features: jsc macros(2) tsconfig
Builtins: "bun:main"
Elapsed: 7ms | User: 7ms | Sys: 5ms
RSS: 23.54MB | Peak: 23.54MB | Commit: 1.07GB | Faults: 45

panic(main thread): Segmentation fault at address 0x2020202020202048
oh no: Bun has crashed. This indicates a bug in Bun, not your code.

To send a redacted crash report to Bun's team,
please file a GitHub issue using the link below:

 https://bun.report/1.1.12/Mr143f0913AiggghE__upk5Tupk5T+lg5Tupk5T2t55T+/t5T+rysR+5trQA2giwgkgBwkwgkgB

zsh: trace trap  bun run index.ts
$ bun run index.ts
1 | import { macroified } from './macro' with { type: "macro" };
    ^
SyntaxError: Unexpected token ':'. Expected a property name.
      at <parse> (/Users/kay/Code/tmp/macrosrepro/index.ts:1:1)

Bun v1.1.12 (macOS arm64)
$ bun run index.ts
1 | import { macroified } from './macro' with { type: "macro" };
2 | console.log(macroified());
    ^
SyntaxError: Unexpected identifier 'YAY'. Expected a ':' following the property name 'static'.
      at <parse> (/Users/kay/Code/tmp/macrosrepro/index.ts:2:1)

Bun v1.1.12 (macOS arm64)
```

### Stack Trace (bun.report)

Bun v1.1.12 ([`43f0913`](<https://github.com/oven-sh/bun/tree/43f0913c3875cbf260c60868268adc5953b09a80>)) on macos aarch64 [RunCommand]

Segmentation fault at address 0x16A9E36A8

- *2 unknown/js code*
- `src.js_parser.NewParser_`
- `src.js_parser.NewParser_`
- `src.js_parser.NewParser_`
- `src.js_parser.NewParser_`
- `src.js_parser.NewParser_`
- `src.js_parser.NewParser_`
- `src.js_parser.NewParser_`
- `src.js_parser.NewParser_`
